taxcalc/parameters.py

```diff
@@ -1,170 +1,496 @@
-                           " values for the years between {0} and {1}.")
-                    raise ValueError(msg.format(self.current_year, year))
-
-            # Set up the '_X = [a, b,...]' variables as 'X = a'
-            self.set_year(self._current_year)
+        if self.current_year != self.start_year:
+            self.set_year(self.start_year)
+        if not reform:
+            return  # no reform to implement
+        reform_years = sorted(list(reform.keys()))
+        last_reform_year = max(reform_years)
+        if last_reform_year > self.end_year:
+            msg = 'reform provision in year={} > end_year={}'
+            ValueError(msg.format(last_reform_year, self.end_year))
+        for year in reform_years:
+            if year != self.start_year:
+                self.set_year(year)
+            self._update({year: reform[year]})
+        self.set_year(self.start_year)
 
     @property
     def current_year(self):
+        """
+        Current policy parameter year property.
+        """
         return self._current_year
 
     @property
     def start_year(self):
+        """
+        First policy parameter year property.
+        """
         return self._start_year
 
     @property
-    def budget_years(self):
-        return self._budget_years
+    def num_years(self):
+        """
+        Number of policy parameter years property.
+        """
+        return self._num_years
 
-    def increment_year(self):
-        self._current_year += 1
-        self.set_year(self._current_year)
+    @property
+    def end_year(self):
+        """
+        Last policy parameter year property.
+        """
+        return self._end_year
 
-    def set_year(self, yr):
-        for name, vals in self._vals.items():
+    def set_year(self, year):
+        """
+        Set policy parameters to values for specified calendar year.
+
+        Parameters
+        ----------
+        year: int
+            calendar year for which to current_year and parameter values
+
+        Raises
+        ------
+        ValueError:
+            if year is not in [start_year, end_year] range.
+
+        Returns
+        -------
+        nothing: void
+
+        Notes
+        -----
+        To increment the current year, use the following statement::
+
+            ppo.set_year(ppo.current_year + 1)
+
+        where ppo is a policy Parameters object.
+        """
+        if year < self.start_year or year > self.end_year:
+            msg = 'year passed to set_year() must be in [{},{}] range.'
+            raise ValueError(msg.format(self.start_year, self.end_year))
+        self._current_year = year
+        year_zero_indexed = year - self._start_year
+        for name in self._vals:
             arr = getattr(self, name)
-            setattr(self, name[1:], arr[yr - self._start_year])
+            setattr(self, name[1:], arr[year_zero_indexed])
+
+    @classmethod
+    def default_data(cls, metadata=False, start_year=None):
```

Comment:
T.J. @talumbau reported this error when using the redesigned Parameters class in the dropQ repository:
```
Traceback (most recent call last):
  File "example.py",
       line 14, in <module>
       mY_dec... = dropq.run_models(tax_dta, num_years=10, user_mods=user_mods)
  File "/Users/talumbau/dev/dropq/dropQ/dropq/dropq.py",
       line 320, in run_models:
       calc3 = calculator(params=params3, records=records3, mods=user_mods)
  File "egg/taxcalc/calculate.py",
       line 63, in calculator [factory function]
  File "egg/taxcalc/calculate.py",
       line 95, in __init__:
       AssertionError
       assert self._params.current_year == self._records.current_year
```
The relevant dropq.run_models code is as follows:
```
def run_models(tax_dta, user_mods="", num_years=NUM_YEARS_DEFAULT,
               start_year=2015):

    # 'fake' advance to handle type changes for jitting
    ...

    #########################################################################
    #   Create Calculators and Masks
    #########################################################################
    records = Records(tax_dta)
    records2 = Records(tax_dta.copy(deep=True))
    records3 = Records(tax_dta.copy(deep=True))
    # add 1 dollar to gross income
    records2.e00200 += 1
    records2.e02000 += 1
    records2.e02615 += 1
    # Default Plans
    # Create a default Parameters object
    params = Parameters(start_year=2013)
    # Create a Calculator
    calc1 = Calculator(params=params, records=records)
    calc1.increment_year()
    calc1.increment_year()
    calc1.calc_all()
    assert calc1.current_year == start_year

    params2 = Parameters(start_year=2013)
    # Create a Calculator with one extra dollar of income
    calc2 = Calculator(params=params2, records=records2)
    calc2.increment_year()
    calc2.increment_year()
    calc2.calc_all()
    assert calc2.current_year == start_year

    # where do the results differ..
    soit1 = results(calc1)
    soit2 = results(calc2)
    mask = (soit1._ospctax != soit2._ospctax)

    # User specified Plans
    params3 = Parameters(start_year=2013)
    # Create a Calculator for the user specified plan
    calc3 = calculator(params=params3, records=records3, mods=user_mods)
    assert calc3.current_year == start_year
    calc3.calc_all()
    soit3 = results(calc3)
    ...
```
Notice that calc1 and calc2, both of which are created using the Calculator constructor, work without error.  The problem comes in using the calculator factory function, which we agreed (at least I thought we agreed) in July to eliminate soon.  One way to think about the error T.J. is getting is that it is caused by an error of mine in the calculator factory function (which is largely mysterious to me).  One strategy is to fix that error, but I'm reluctant to put in effort on something that is slated for extinction.  The other strategy is to simply use the new Parameters class the way it is supposed to be used and rewrite the calc3 part of the dropq.run_models code as follows:
```
    # User specified Plans
    params3 = Parameters(start_year=2013)
    params3.implement_reform(user_mods)
    # Create a Calculator for the user specified plan
    calc3 = Calculator(params=params3, records=records3)
    calc3.increment_year()
    calc3.increment_year()
    calc3.calc_all()
    assert calc3.current_year == start_year
```
Does this code produce the same error as T.J. reported above?  I'm asking because I do not have the knowledge or capabilities to run the webapp on my computer.